import {Paper, SvgIcon} from '@material-ui/core';
import Box from '@material-ui/core/Box';
import Link from '@material-ui/core/Link';
import Typography from '@material-ui/core/Typography';
import ReactMarkdown from 'markdown-to-jsx';
import React from 'react';
import {connect} from 'react-redux';
import {reactMarkdownOptions} from './util';

function LandingPage(props) {
    return <Paper elevation={0}>

        <Typography variant="h5"><SvgIcon style={{verticalAlign: 'text-bottom', paddingBottom: 3}}>
            <path
                d="M19.21 12.04l-1.53-.11-.3-1.5C16.88 7.86 14.62 6 12 6 9.94 6 8.08 7.14 7.12 8.96l-.5.95-1.07.11C3.53 10.24 2 11.95 2 14c0 2.21 1.79 4 4 4h13c1.65 0 3-1.35 3-3 0-1.55-1.22-2.86-2.79-2.96z"
                fill="#4285F4"></path>
            <path
                d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM19 18H6c-2.21 0-4-1.79-4-4 0-2.05 1.53-3.76 3.56-3.97l1.07-.11.5-.95C8.08 7.14 9.94 6 12 6c2.62 0 4.88 1.86 5.39 4.43l.3 1.5 1.53.11c1.56.1 2.78 1.41 2.78 2.96 0 1.65-1.35 3-3 3z"></path>
        </SvgIcon> Cirrocumulus</Typography>
        <p>Cirrocumulus is an interactive visualization tool for large-scale single-cell genomics
            data, with the following key features:</p>
        <ul>
            <li>Run on a laptop, on-premise server, cloud VM, or Google App Engine</li>
            <li>View spatial transcriptomics data overlaid on an image</li>
            <li>Share the current visualization state in a URL</li>
            <li>Share datasets securely with collaborators</li>
            <li>Create dotplots, heatmaps, and violin plots to explore relationships between categorical variables and
                expression
            </li>
            <li>Explore complete differential expression results generated by <Link target="_blank"
                                                                                    rel="noopener noreferrer"
                                                                                    href="https://scanpy.readthedocs.io/">Scanpy</Link> or <Link
                target="_blank" rel="noopener noreferrer"
                href="https://pegasus.readthedocs.io/">Pegasus</Link>/<Link target="_blank"
                                                                            rel="noopener noreferrer"
                                                                            href="https://cumulus.readthedocs.io/">Cumulus</Link>
            </li>
            <li>Interactive differential expression analysis (Mann-Whitney U test)</li>
            <li>Create and share “AND” or “OR” filters</li>
            <li>Collaboratively annotate cell types in real time</li>
            <li>Quickly load multiple features from predefined lists</li>
            <li>Explore multiple features and embeddings simultaneously</li>
            <li>Fast exploration of 2 and 3-d embeddings of millions of cells, including zoom, pan, rotate
                (3-d), and lasso tools
            </li>
            <li>Save publication quality images</li>
            <li>Highly customizable - for example, set the color map, point size, or whether to use fog for 3-d
                embeddings to fade distant points
            </li>
            <li>Visualize datasets in <Link target="_blank" rel="noopener noreferrer"
                                            href="https://anndata.readthedocs.io/">h5ad</Link>, <Link target="_blank"
                                                                                                      rel="noopener noreferrer"
                                                                                                      href="https://linnarssonlab.org/loompy/format/">loom</Link>, <Link
                target="_blank" rel="noopener noreferrer" href="https://satijalab.org/seurat/">Seurat</Link>, or <Link
                target="_blank" rel="noopener noreferrer"
                href="https://github.com/STAR-Fusion/STAR-Fusion/wiki">STAR-Fusion</Link> formats
            </li>
        </ul>

        <Typography variant="h5">Links</Typography>

        <ul>
            <li><Link target="_blank" rel="noopener noreferrer"
                      href="http://cirrocumulus.readthedocs.io/">Documentation</Link></li>
            <li>Contact: cirrocumulus@broadinstitute.org</li>
            <li><Link target="_blank" rel="noopener noreferrer"
                      href="https://github.com/klarman-cell-observatory/cirrocumulus">Source
                Code</Link></li>
        </ul>

        <Typography variant="h5">Primary Embedding Controls</Typography>
        <ul>
            <li>Pan: Mouse left click (2-d), right click (3-d)</li>
            <li>Rotate 3-d: Mouse left click</li>
            <li>Zoom: Mouse wheel</li>
            <li>Select: When using lasso or select tool, hold down the Ctrl or Command key to add to selection</li>
            <li>Resize: Click and drag the divider below the primary embedding</li>
        </ul>

        <Typography variant="h5">Embedding Gallery</Typography>
        <ul>
            <li>Drag charts to reorder</li>
            <li>Click chart to set primary view</li>
        </ul>
        {props.serverInfo && props.serverInfo.footer &&
        <Box><ReactMarkdown options={reactMarkdownOptions} children={props.serverInfo.footer}/></Box>}
    </Paper>;
}


const mapStateToProps = state => {
    return {
        serverInfo: state.serverInfo
    };
};
const mapDispatchToProps = dispatch => {
    return {};
};

export default (connect(
    mapStateToProps, mapDispatchToProps,
)(LandingPage));
